### ISSUE TITLE
Pylance crashing for unknown reason

### ISSUE THREAD

**Issue by JesseNoEyes** [microsoft/pylance-release#4712]
## Environment data

-   Language Server version: v2023.8.20
-   OS and version: Windows 10.0.19045 Build 19045
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.11.4

## Code Snippet

```python
import pygame as pg
.
.
.                    
            for event in pg.event.get():    
                if event.type == pg.MOUSEBUTTONDOWN:
                    if self.shift:
                        if event.button == 4:
                            self.tile_variant = (self.tile_variant - 1) % len(
                                self.assets[self.tile_list[self.tile_group]]
                            )
                        if event.button == 5:
                            self.tile_variant = (self.tile_variant + 1) % len(
                                self.assets[self.tile_list[self.tile_group]]
                            )

                    else:
                        if event.button == 4:
                            self.tile_group = (self.tile_group - 1) % len(
                                self.tile_list
                            )
                            self.tile_variant = 0
                        elif event.button == 5:
                            self.tile_group = (self.tile_group + 1) % len(
                                self.tile_list
                            )
                            self.tile_variant = 0
```

## Repro Steps

After getting the included error and noticing that Pylance was breaking throughout the code, I commented out sections until I pinpointed the above code as the issue. Specifically, the last "else" statement. The Black formatter splits the "self.tile_group..." line that way, but the error persists even when it is in a single line. It should be noted that the code still executes and works fine, but Pylance breaks throughout the entire document.

## Logs

```
[{
	"resource": "~/coding workspace/python workspace/platformer_new/editor.py",
	"owner": "_generated_diagnostic_collection_name_#3",
	"severity": 8,
	"message": "An internal error occurred while type checking file \"~\\python workspace\\platformer_new\\editor.py\": Error: Debug Failure. setIncompleteSubtype can be called only on a valid incomplete cache entry\n    at w (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:176588)\n    at ~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:184022\n    at Array.forEach (<anonymous>)\n    at F (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:183309)\n    at D (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:180170)\n    at Object.getTypeFromCodeFlow (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:184809)\n    at ua (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:746250)\n    at ~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:535128\n    at Me (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:537480)\n    at Ke (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:567327)\n    at En (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:703895)\n    at pa (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:744984)\n    at ~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:763761\n    at ca (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:745627)\n    at Aa (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:763745)\n    at ~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:767104\n    at Array.forEach (<anonymous>)\n    at Na (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:766551)\n    at Ra (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:772401)\n    at It (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:598454)\n    at Be (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:565438)\n    at _t (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:593931)\n    at _t (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:593601)\n    at ~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:537662\n    at Me (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:538322)\n    at oa (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:744015)\n    at ~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:532527\n    at ca (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:745627)\n    at Fe (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:532516)\n    at a.timeOperation (c:\\Users\\Sky\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:1090716)\n    at R._validateConditionalIsBool (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:98042)\n    at R.visitIf (~\\.vscode\\extensions\\ms-python.vscode-pylance-2023.8.20\\dist\\pyright.bundle.js:1:84944)",
	"source": "Pylance",
	"startLineNumber": 1,
	"startColumn": 1,
	"endLineNumber": 1,
	"endColumn": 1
}]
```

**Comment by erictraut** [comment]
Could you provide a self-contained minimal code sample that demonstrates the problem? I'm not able to repro the problem you're seeing with the code above, but that's likely because it's incomplete. It references symbols that are not defined.

**Comment by erictraut** [comment]
Based on the error message, this appears to be the same as [this issue](https://github.com/microsoft/pyright/issues/5622). We've seen evidence of this bug in the pylance telemetry, but we didn't have sufficient information to repro the issue or diagnose its root cause. So I'd really appreciate any additional details that could help us track down the problem!

**Comment by JesseNoEyes** [comment]
here is the entire object.
'''
# imports
import pygame as pg
import sys

from scripts.tilemap import Tilemap
from scripts.utils import load_images, load_image

RENDER_SCALE = 2.0

pg.init()

WIDTH, HEIGHT = 1024, 768
GREY = (128, 128, 128)
BLACK = (0, 0, 0)


# editor object
class Editor:
    def __init__(self):
        # window and display surface setup
        self.win = pg.display.set_mode((WIDTH, HEIGHT))
        self.display = pg.Surface((WIDTH / 2, HEIGHT / 2))
        pg.display.set_caption("editor")

        # loading assets
        self.assets = {
            "decor": load_images("tiles/decor"),
            "grass": load_images("tiles/grass"),
            "stone": load_images("tiles/stone"),
            "large_decor": load_images("tiles/large_decor"),
        }

        # attributes
        self.clock = pg.time.Clock()
        self.tilemap = Tilemap(self)
        self.draw_grid_lines = False
        self.scroll = [0, 0]
        self.movement = [False, False, False, False]
        self.tile_list = list(self.assets)
        self.tile_group = 0
        self.tile_variant = 0
        self.clicking = False
        self.right_clicking = False
        self.middle_clicking = False
        self.shift = False
        self.on_grid = True

    # rendering
    def draw_window(self, tile_pos, current_tile_img, render_scroll, on_grid, mousepos):
        self.win.blit(pg.transform.scale(self.display, self.win.get_size()), (0, 0))
        self.display.fill(BLACK)

        self.tilemap.render(self.display, offset=render_scroll)

        if on_grid:
            self.display.blit(
                current_tile_img,
                (
                    tile_pos[0] * self.tilemap.tile_size - self.scroll[0],
                    tile_pos[1] * self.tilemap.tile_size - self.scroll[1],
                ),
            )
        else:
            self.display.blit(current_tile_img, mousepos)

        if self.draw_grid_lines:
            for i in range(WIDTH // self.tilemap.tile_size):
                pg.draw.line(
                    self.display,
                    GREY,
                    (i * self.tilemap.tile_size, 0),
                    (i * self.tilemap.tile_size, HEIGHT),
                )
            for i in range(HEIGHT // self.tilemap.tile_size):
                pg.draw.line(
                    self.display,
                    GREY,
                    (0, i * self.tilemap.tile_size),
                    (WIDTH, i * self.tilemap.tile_size),
                )

    # main loop
    def run(self):
        run = True

        while run == True:
            self.clock.tick(60)

            mouse_pos = pg.mouse.get_pos()
            mouse_pos = (mouse_pos[0] / RENDER_SCALE, mouse_pos[1] / RENDER_SCALE)
            tile_pos = (
                int((mouse_pos[0] + self.scroll[0]) // self.tilemap.tile_size),
                int((mouse_pos[1] + self.scroll[1]) // self.tilemap.tile_size),
            )

            self.scroll[0] += (self.movement[0] - self.movement[1]) * 2
            self.scroll[1] += (self.movement[3] - self.movement[2]) * 2
            render_scroll = (int(self.scroll[0]), int(self.scroll[1]))

            if self.clicking:
                if self.on_grid:
                    self.tilemap.tilemap[f"{tile_pos[0]};{tile_pos[1]}"] = {
                        "type": self.tile_list[self.tile_group],
                        "variant": self.tile_variant,
                        "pos": tile_pos,
                    }
                else:
                    self.tilemap.offgrid_tiles.append(
                        {
                            "type": self.tile_list[self.tile_group],
                            "variant": self.tile_variant,
                            "pos": (
                                mouse_pos[0] + self.scroll[0],
                                mouse_pos[1] + self.scroll[1],
                            ),
                        }
                    )

            if self.right_clicking:
                tile_loc = f"{tile_pos[0]};{tile_pos[1]}"
                if tile_loc in self.tilemap.tilemap:
                    del self.tilemap.tilemap[tile_loc]
                for tile in self.tilemap.offgrid_tiles.copy():
                    tile_img = self.assets[tile["type"]][tile["variant"]]
                    tile_r = pg.Rect(
                        tile["pos"][0] - self.scroll[0],
                        tile["pos"][1] - self.scroll[1],
                        tile_img.get_width(),
                        tile_img.get_height(),
                    )
                    if tile_r.collidepoint(mouse_pos):
                        self.tilemap.offgrid_tiles.remove(tile)

            if self.middle_clicking:
                tile_loc = f"{tile_pos[0]};{tile_pos[1]}"
                if tile_loc in self.tilemap.tilemap:
                    tile_type = self.tilemap.tilemap[tile_loc]["type"]
                    if tile_type == "decor":
                        tile_type_int = 0
                    elif tile_type == "grass":
                        tile_type_int = 1
                    elif tile_type == "stone":
                        tile_type_int = 2
                    elif tile_type == "large_decor":
                        tile_type_int = 3

                    self.tile_group = tile_type_int
                    self.tile_variant = self.tilemap.tilemap[tile_loc]["variant"]

            current_tile_img = self.assets[self.tile_list[self.tile_group]][
                self.tile_variant
            ].copy()
            current_tile_img.set_alpha(100)

            # event checks
            for event in pg.event.get():
                # allow exit
                if event.type == pg.QUIT:
                    run = False

                # movement
                if event.type == pg.KEYDOWN:
                    if event.key == pg.K_d:
                        self.movement[0] = True
                    if event.key == pg.K_a:
                        self.movement[1] = True
                    if event.key == pg.K_w:
                        self.movement[2] = True
                    if event.key == pg.K_s:
                        self.movement[3] = True
                    if event.key == pg.K_o:
                        self.on_grid = not self.on_grid

                    # draw grid lines toggle
                    if event.key == pg.K_g:
                        if self.draw_grid_lines:
                            self.draw_grid_lines = False
                        else:
                            self.draw_grid_lines = True
                    if event.key == pg.K_LSHIFT:
                        self.shift = True

                if event.type == pg.MOUSEBUTTONDOWN:
                    if event.button == 1:
                        self.clicking = True
                    if event.button == 2:
                        self.middle_clicking = True
                    if event.button == 3:
                        self.right_clicking = True

                    if self.shift:
                        if event.button == 4:
                            self.tile_variant = (self.tile_variant - 1) % len(
                                self.assets[self.tile_list[self.tile_group]]
                            )
                        if event.button == 5:
                            self.tile_variant = (self.tile_variant + 1) % len(
                                self.assets[self.tile_list[self.tile_group]]
                            )

                    else:
                        if event.button == 4:
                            self.tile_group = (self.tile_group - 1) % len(
                                self.tile_list
                            )
                            self.tile_variant = 0
                        elif event.button == 5:
                            self.tile_group = (self.tile_group + 1) % len(
                                self.tile_list
                            )
                            self.tile_variant = 0

                # stop movement on keyup
                if event.type == pg.KEYUP:
                    if event.key == pg.K_d:
                        self.movement[0] = False
                    if event.key == pg.K_a:
                        self.movement[1] = False
                    if event.key == pg.K_w:
                        self.movement[2] = False
                    if event.key == pg.K_s:
                        self.movement[3] = False
                    if event.key == pg.K_LSHIFT:
                        self.shift = False

                if event.type == pg.MOUSEBUTTONUP:
                    if event.button == 1:
                        self.clicking = False
                    if event.button == 2:
                        self.middle_clicking = False
                    if event.button == 3:
                        self.right_clicking = False

            self.draw_window(
                tile_pos, current_tile_img, render_scroll, self.on_grid, mouse_pos
            )

            pg.display.flip()

        pg.quit()
        sys.exit()


if __name__ == "__main__":
    Editor().run()
'''
and here is the tilemap script it imports, if that is important.
'''
import pygame as pg

NEIGHBOR_OFFSETS = [
    (-1, 0),
    (-1, -1),
    (0, -1),
    (0, 0),
    (1, -1),
    (1, 1),
    (1, 0),
    (0, 1),
    (-1, 1),
]
PHYSICS_TILES = {"grass", "stone"}


class Tilemap:
    def __init__(self, game, tile_size=16):
        self.game = game
        self.tile_size = tile_size
        self.tilemap = {}
        self.offgrid_tiles = []

    def tiles_around(self, pos):
        tiles = []
        tile_loc = (int(pos[0] // self.tile_size), int(pos[1] // self.tile_size))
        for offset in NEIGHBOR_OFFSETS:
            check_loc = f"{tile_loc[0] + offset[0]};{tile_loc[1] + offset[1]}"
            if check_loc in self.tilemap:
                tiles.append(self.tilemap[check_loc])
        return tiles

    def physics_tiles_around(self, pos):
        rects = []
        for tile in self.tiles_around(pos):
            if tile["type"] in PHYSICS_TILES:
                rects.append(
                    pg.Rect(
                        tile["pos"][0] * self.tile_size,
                        tile["pos"][1] * self.tile_size,
                        self.tile_size,
                        self.tile_size,
                    )
                )
        return rects

    def render(self, surface, offset=(0, 0)):
        for tile in self.offgrid_tiles:
            surface.blit(
                self.game.assets[tile["type"]][tile["variant"]],
                (tile["pos"][0] - offset[0], tile["pos"][1] - offset[1]),
            )

        for x in range(
            offset[0] // self.tile_size,
            (offset[0] + surface.get_width()) // self.tile_size + 1,
        ):
            for y in range(
                offset[1] // self.tile_size,
                (offset[1] + surface.get_height()) // self.tile_size + 1,
            ):
                loc = f"{x};{y}"
                if loc in self.tilemap:
                    tile = self.tilemap[loc]
                    surface.blit(
                        self.game.assets[tile["type"]][tile["variant"]],
                        (
                            tile["pos"][0] * self.tile_size - offset[0],
                            tile["pos"][1] * self.tile_size - offset[1],
                        ),
                    )

'''

**Comment by erictraut** [comment]
And what is `scripts`? Is that a public library? An internal module?

What would really help is if you can produce a self-contained minimal code sample that demonstrates the problem. The above code is neither self-contained nor minimal.

**Comment by JesseNoEyes** [comment]
scripts is just a local module i made containing two functions to load images from a provided directory. i can provide its code if you'd like, but that would make it even less minimal.

i do not know how to produce such a self-contained, minimal code sample as i do not know what is causing the issue to begin with. even after copious amounts of toying and rearranging and commenting, all i could ascertain is that the above code in its entirety produces the error i provided. when the "else" statement i first provided is commented out, the error goes away. in isolation, it doesn't produce the error.

**Comment by erictraut** [comment]
I appreciate your help, but I'm not able to repro the problem with the code you've provided above. If you're willing to do so, could you create a new project with a single file `test.py` and copy the above code into it. Does this repro the problem for you? If not, see what else you need to copy from your first project to get it to occur.

**Comment by erictraut** [comment]
Here's another possible plan of attack. In an effort to get additional insights into this bug, I recently added more detailed logging. This change will be in next week's pylance release, but it's already in this week's prerelease build. If you're willing to switch (temporarily) to the prerelease build of pylance and repro the problem on your machine, it would be helpful to see what error is logged. To install the prerelease version, just go to the Extensions panel in VS Code, find Pylance and right click on the gear icon. Then choose "Switch to Pre-Release Version".

**Comment by JesseNoEyes** [comment]
An internal error occurred while type checking file "~\coding workspace\python workspace\platformer_new\editor.py": Error: Debug Failure. setIncompleteSubtype can be called only on a valid incomplete cache entry: prev cache entry?: false index=0 isPending=falseevaluationCount=1
    at w (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:176648)
    at ~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:184154
    at Array.forEach (<anonymous>)
    at F (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:183441)
    at D (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:180302)
    at Object.getTypeFromCodeFlow (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:184941)
    at ua (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:748089)
    at ~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:536213
    at Me (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:538932)
    at Ke (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:568807)
    at En (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:705528)
    at pa (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:746823)
    at ~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:765607
    at ca (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:747466)
    at Aa (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:765591)
    at ~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:768953
    at Array.forEach (<anonymous>)
    at Na (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:768400)
    at Ra (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:773922)
    at It (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:599993)
    at Be (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:566918)
    at _t (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:595470)
    at _t (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:595140)
    at ~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:539121
    at Me (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:539781)
    at oa (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:745854)
    at ~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:533612
    at ca (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:747466)
    at Fe (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:533601)
    at a.timeOperation (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:1094168)
    at R._validateConditionalIsBool (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:98098)
    at R.visitIf (~\.vscode\extensions\ms-python.vscode-pylance-2023.8.21\dist\pyright.bundle.js:1:85000)

**Comment by JesseNoEyes** [comment]
I one by one pasted the other modules all into the same `test.py` file. After each step and at the end, the error is still thrown.

**Comment by JesseNoEyes** [comment]
I also just noticed that now that all the code is in one long file, commenting out the offending else statement no longer resolves the issue as it did previously.

**Comment by erictraut** [comment]
When you say "pasted the other modules all into the same", which modules are you referring to?

Are you able to repro the crash with only the source code you pasted above? (Presumably, you'll see errors about the `script` import failing in that case.)

Maybe it's dependent on the source in the pygame library. Which version of that library do you have installed in your python environment?

**Comment by JesseNoEyes** [comment]
I'm referring to the `scripts.utils` and `tilemap` modules i import.
just the code above alone, I get an import error. "No module 'scripts.utils' found". If I comment out the imports themselves, I just get errors about functions not being defined in my code due to the lack of imports.
the version of pygame i am using is the community edition, pygame-ce 2.3.1

**Comment by JesseNoEyes** [comment]
To be thorough, I just tested, and the error is still present when I have only the regular pygame installed.

**Comment by erictraut** [comment]
I just switched to pygame-ce==2.3.1, and I can now repro the problem! Thanks for your help. I should be able to take it from here.

**Comment by erictraut** [comment]
This was a subtle bug, but it was easy to fix once I understood the problem. It will be fixed in the next release.

@rchiodo, this is the [same issue](https://github.com/microsoft/pyright/issues/5622) that you reported seeing in the pylance telemetry. It would be good to get this into next week's production release of pylance. It could explain some of the various reports of crashes in notebooks and cases where semantic token highlighting stops working. Would it make sense to spin a hot fix for the prerelease version that cherry picks this fix? Here's the commit: https://github.com/microsoft/pyright/commit/bf4749809e2a2f580184a1cacd44bc2c37e52b4f.

I wasn't able to come up with a good unit test for this condition because it occurs only in really rare cases involving complex code flow graphs with loops.

**Comment by debonte** [comment]
This issue has been fixed in prerelease version 2023.8.31, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2023831-16-August-2023-PreRelease)
